File: scripts/Guard.lua
---@uuid 9f8183d3-2a4a-4951-a6e6-5fbc9c479e2e
---@class Guard : Script
---@field private reloading_sec number
---@field reload_delay_sec number
---@field gun_height number
---@field switch_waypoint_timeout_sec number
---@field private waypoint_sec number
---@field private current_waypoint Vector3
---@field private collider Node
---@field bullet_prefab Prefab
---@field initial_bullet_velocity number
---@field attack_range number
---@field beacon_reached_distance number
---@field move_power number
Guard = script_class()

---@param ctx ScriptContext
---@return boolean
function Guard:try_attack_player(ctx)
    local player_pos = ctx.plugins:get("Game").player:global_position()
    local self_pos = ctx.handle:global_position()
    local sight_vector = player_pos:sub(self_pos)
    print(string.format(
        "try attack player. player_pos: %s, self_pos: %s",
        player_pos, self_pos
    ))

    if self:can_see_player(ctx, player_pos, sight_vector) then
        Bullet:spawn(
            ctx.scene,
            {
                prefab = self.bullet_prefab,
                origin = self_pos + Vector3:new(0.0, self.gun_height, 0.0),
                direction = sight_vector,
                initial_velocity = self.initial_bullet_velocity,
                author_collider = self.collider,
                range = self.attack_range,
            }
        )
        self.reloading_sec = self.reload_delay_sec
        return true
    end

    return false
end

---@param ctx ScriptContext
---@param player_pos Vector3
---@param sight_vector Vector3
---@return boolean
function Guard:can_see_player(ctx, player_pos, sight_vector)
    ---@type RayCastOptions
    local opts = {
        ray_origin = Point3:from(player_pos),
        ray_direction = sight_vector:normalize(),
        max_len = sight_vector:magnitude(),
        sort_results = true
    }
    local results = {}
    physics:cast_ray(ctx, opts, results)
    for i, hit in ipairs(results) do
        local node = hit.collider
        if node ~= self.collider then
            while node:is_some() do
                if node:has_script("Player") then
                    return true
                end
                node = node:parent()
            end
            return false
        end
    end
    return false
end

---@param ctx ScriptContext
function Guard:move_to_waypoint(ctx)
    self.waypoint_sec = self.waypoint_sec + ctx.dt
    if self.waypoint_sec > self.switch_waypoint_timeout_sec then
        self.current_waypoint = nil;
        self.waypoint_sec = 0.0;

        -- println!("guard {:?end switched waypoint", ctx.handle);
    end
    if self.current_waypoint ~= nil then
        local beacons = ctx.plugins:get("Game").beacons;
        self.current_waypoint = beacons[math.random(#beacons)]
    end
    local pos = ctx.handle
        :local_transform()
        :position()
        :get_value_ref();
    local vector_to_beacon = self.current_waypoint.unwrap().sub(pos)
    if vector_to_beacon:magnitude() < self.beacon_reached_distance then
        self.current_waypoint = nil;
    else
        local force = vector_to_beacon.normalize() * self.move_power;
        ctx.handle
            :as_rigid_body_mut()
            :unwrap()
            :apply_force(force);
    end
end


---@param ctx ScriptContext
function Guard:on_init(ctx)
    self.collider = ctx
        .handle
        .try_get_collider(ctx.scene)
        :expect("Collider not found under Guard node");
end

function Guard:on_start()
    ctx.message_dispatcher.subscribe_to("BulletHit", ctx.handle);
end

---@param ctx ScriptContext
function Guard:on_update(ctx)
    if self.reloading_sec > 0.0 then
        self.reloading_sec = self.reloading_sec - ctx.dt;
    end

    if self.reloading_sec > 0.0 or not self:try_attack_player(ctx) then
        self:move_to_waypoint(ctx)
    end
end

---@param ctx ScriptContext
---@param message ScriptMessagePayload
function Guard:on_message(ctx, message)
    local bullet = message:downcast_ref("BulletHit")
    if bullet then
        ctx.handle:remove_node();
        ctx.plugins:get_mut("Game"):inc_frags()
        print("guard killed!")
    end
end

File: cs/examples/Guards/Guard.cs
using System;
using System.Collections.Generic;
using FyroxLite;

[Uuid("9f8183d3-2a4a-4951-a6e6-5fbc9c479e2e")]
public class Guard : NodeScript
{
    private float reload_delay_sec;
    private float gun_height;
    private float switch_waypoint_timeout_sec;
    private Prefab bullet_prefab;
    private float initial_bullet_velocity;
    private float attack_range;
    private float beacon_reached_distance;
    private float move_power;
 
    [HideInInspector] [Transient] private float reloading_sec;   
    [HideInInspector] [Transient] private float waypoint_sec;
    [HideInInspector] [Transient] private Vector3? current_waypoint;
    [HideInInspector] [Transient] private Node collider;
    [HideInInspector] [Transient] private int id;

    private const int FRACTION_GUARDS = 1;

    public void Init(int id)
    {
        this.id = id;
    }

    private bool TryAttackPlayer()
    {
        var playerPos = Plugin.Get<Game>().player.GlobalPosition;
        var selfPos = Node.GlobalPosition;
        var sightVector = playerPos - selfPos;

        if (CanSeePlayer(playerPos, sightVector))
        {
            Bullet.Spawn(new Bullet.BulletSeed
            {
                Prefab = bullet_prefab,
                Origin = selfPos + Vector3.Up * gun_height,
                Direction = sightVector,
                InitialVelocity = initial_bullet_velocity,
                AuthorCollider = collider,
                Range = attack_range,
                Fraction = FRACTION_GUARDS
            });
            reloading_sec = reload_delay_sec;
            return true;
        }

        return false;
    }

    private bool CanSeePlayer(Vector3 playerPos, Vector3 sightVector)
    {
        var results = Physics.CastRay(new RayCastOptions
        {
            RayOrigin = playerPos,
            RayDirection = sightVector.Normalized(),
            MaxLen = sightVector.Length(),
            SortResults = true
        });
        foreach (var hit in results)
        {
            var node = hit.Collider;
            if (node != collider)
            {
                while (node.Alive)
                {
                    if (node.FindScript<Player>() != null)
                    {
                        return true;
                    }

                    node = node.Parent;
                }

                return false;
            }
        }

        return false;
    }

    private void MoveToWaypoint(float dt)
    {
        waypoint_sec += dt;
        if (waypoint_sec > switch_waypoint_timeout_sec)
        {
            current_waypoint = null;
            waypoint_sec = 0.0f;
        }

        if (current_waypoint == null)
        {
            var beacons = Plugin.Get<Game>().beacons;
            current_waypoint = beacons[new Random().Next(beacons.Count)];
        }

        var vectorToBeacon = current_waypoint.Value - Node.LocalPosition;
        if (vectorToBeacon.Length() < beacon_reached_distance)
        {
            current_waypoint = null;
        }
        else
        {
            Node.AsRigidBody().Value.ApplyForce(vectorToBeacon.Normalized() * move_power);
        }
    }

    protected override void OnInit()
    {
        collider = Node.FindColliderInChildren() ?? throw new Exception("Collider not found under Guard node");
    }

    protected override void OnStart()
    {
        Node.SubscribeTo();
    }

    protected override void OnUpdate(float dt)
    {
        if (reloading_sec > 0.0f)
        {
            reloading_sec -= dt;
        }

        if (reloading_sec > 0.0f || !TryAttackPlayer())
        {
            MoveToWaypoint(dt);
        }
    }

    protected override void OnMessage(object message)
    {
        if (message is BulletHitMessage hit && hit.Fraction != FRACTION_GUARDS)
        {
            Node.Destroy();
            Plugin.Get<Game>().IncFrags();
            Console.WriteLine("Guard killed!");
        }
    }
}